### PR TITLE
Custom Event handling methods as solution to circular dependencies

### PR DESCRIPTION
**Pull request by bozdoz**
Add events to deal with circular dependencies; added some type annotations and definitions; cleaned up some methods.

Also, events could make for a nicer/more flexible API, going forward.

Think of this line:

```js
  // signal to all listeners that introjs has exited
  this.fire('exit');

  // check if any callback is defined
  if (this._introExitCallback !== undefined) {
    this._introExitCallback.call(this);
  }
```

Devs can add all the introJs.on('exit') handlers they want.

### CODE REVIEW

File: src/core/exitIntro.js [hunk 1]
```diff
@@ -34,15 +32,6 @@ export default function exitIntro(targetElement, force) {
     forEach(overlayLayers, (overlayLayer) => removeChild(overlayLayer));
   }
 
-  //remove all helper layers
-  const helperLayer = targetElement.querySelector(".introjs-helperLayer");
```
Comment (afshinm) [inline src/core/exitIntro.js:38]: Why we don't need this anymore?

File: src/core/exitIntro.js [hunk 2]
```diff
@@ -55,15 +44,13 @@ export default function exitIntro(targetElement, force) {
 
   removeShowElement();
 
-  //clean listeners
-  DOMEvent.off(window, "keydown", onKeyDown, this, true);
-  DOMEvent.off(window, "resize", onResize, this, true);
```
Comment (afshinm) [inline src/core/exitIntro.js:60]: Are we handling these somewhere else?

File: src/core/exitIntro.js [hunk 3]
```diff
@@ -55,15 +44,13 @@ export default function exitIntro(targetElement, force) {
...
+  // signal to all listeners that introjs has exited
+  this.fire('exit');
```
Comment (afshinm) [inline src/core/exitIntro.js:48]: :100:

File: src/core/exitIntro.js (hunk 2 above)
Comment (afshinm) [reply src/core/exitIntro.js:60]: Oh nvm, just saw the `off` handler. nice

**Review by afshinm — COMMENTED** [review]
I really like this idea but I believe we should still support the `introjs(...).oncomplete(fn => ...)` notion as well (in addition to `introjs(...).of('complete', fn => ...)`. I don't really want to introduce a massive backward incompatibility. But I can definitely see that this approach can simplify the codebase.

File: src/core/exitIntro.js (hunk 1 above)
Comment (bozdoz) [reply src/core/exitIntro.js:38]: removed in another file

**Comment by bozdoz** [comment]
> I really like this idea but I believe we should still support the `introjs(...).oncomplete(fn => ...)` notion as well (in addition to `introjs(...).of('complete', fn => ...)`. I don't really want to introduce a massive backward incompatibility. But I can definitely see that this approach can simplify the codebase.

I'll take a look at `oncomplete`.  Did I remove any of that functionality?

**Comment by bozdoz** [comment]
Just noticing that exitIntro is always called after the oncomplete:

```js
      if (
        this._introItems.length - 1 === this._currentStep &&
        typeof this._introCompleteCallback === "function"
      ) {
        this._introCompleteCallback.call(this);
      }

      exitIntro.call(this, this._targetElement);
```

maybe we can just put that at the top of exitIntro instead

**Comment by afshinm** [comment]
> I'll take a look at oncomplete. Did I remove any of that functionality?

Sorry no, I sort of assumed the long term plan is to replace the new event handle (`intro.on(...)`) with the callbacks (`intro.on...()`)?

> maybe we can just put that at the top of exitIntro instead

That makes sense @bozdoz.

**Comment by bozdoz** [comment]
> Just noticing that exitIntro is always called after the oncomplete:
> 
> ```js
>       if (
>         this._introItems.length - 1 === this._currentStep &&
>         typeof this._introCompleteCallback === "function"
>       ) {
>         this._introCompleteCallback.call(this);
>       }
> 
>       exitIntro.call(this, this._targetElement);
> ```
> 
> maybe we can just put that at the top of exitIntro instead

I was wrong about this. It's called if skip button is pressed on the last tip, and if nextStep is called when there is no nextStep.  Seems fair to keep it.

**Pull request closed without merging**